Feat:add linux arm image to CI

TG199 opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #35023

<!-- Either: -->
- [x] These changes do not require tests because they are CI

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

TG199 committed:
chore: add linux-arm.yml workflow for Linux ARM builds

This commit introduces a new GitHub Actions workflow file (linux-arm.yml) for
building and testing Servo on an ARM runner (ubuntu-24.04-arm). The workflow sets up
the environment, builds Servo using an ARM target, and runs unit tests and bencher
jobs when enabled. Note that WPT tests are omitted for ARM as specified.

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
chore: add linux-arm job to dispatch-workflow.yml

This commit updates dispatch-workflow.yml to include a new job entry for Linux ARM.
When the matrix contains linux-arm as the workflow, it now triggers the linux-arm.yml
workflow. This change allows ARM builds to be dispatched as part of the CI pipeline.

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
feat(try-parser): add linux-arm support to try_parser.py

Extend try_parser.py to recognize the "linux-arm" keyword. When detected, it now returns
a JobConfig with the workflow set to "linux-arm" and an updated name ("Linux ARM"). This
ensures that the generated matrix includes an ARM job when requested.

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
Dummy change, will revert later

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
reverting dummy change

Signed-off-by: TG <ebiritg@gmail.com>

TG199 commented:
Hi @sagudev,

I've been working on adding Linux ARM support to Servo CI. My approach was to:

- Create a new workflow file (linux-arm.yml) that mirrors the existing Linux workflow but uses the GitHub-hosted ARM runner (ubuntu-24.04-arm) and excludes WPT tests.

- Update dispatch-workflow.yml to add a new job for "linux-arm" so that when the try parser outputs that workflow, it triggers the ARM job.

- Modify try_parser.py to recognize keywords like "linux-arm" and return a JobConfig with workflow set to "linux-arm" and an updated name (e.g., "Linux ARM").

When I manually triggered the workflow using the updated configuration, I encountered an error:

Invalid workflow file: too many workflows are referenced, total: 21, limit: 20

According to the try_parser output, I expected a matrix with 6 entries (for linux, macos, windows, android, ohos, and lint). It seems the matrix expansion in dispatch-workflow.yml is resulting in 21 workflow calls.

Could you please advise on how to adjust the configuration or matrix to avoid exceeding the workflow reference limit? Thanks for your help!

sagudev commented:
> Invalid workflow file: too many workflows are referenced, total: 21, limit: 20

Hm, looks like this is undocumented limitation: https://github.com/orgs/community/discussions/32192, let's keep `linux-arm` out of `full` for now, so we would only support running `mach try linux-arm` for now. We will need to design different architecture to bypass this limitation (I think eliminating dispatch-workflow would be enough).

TG199 committed:
Remove duplicate check on linux-arm

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
Remove linux arm in dispatch workflow

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
Change logic on linux-arm check

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
Add linux-arm to dispatch workflow

Signed-off-by: TG <ebiritg@gmail.com>

TG199 commented:
Hi @sagudev,

I’ve updated the try_parser to filter out the linux-arm preset when running a full run. In the handle_preset method and in the full-run section of Config.parse(), I attempted to ensure that linux-arm isn’t included in the matrix for mach try full. However, I'm still seeing the same error in GitHub Actions:

Invalid workflow file: .github/workflows/try.yml: too many workflows are referenced, total: 21, limit: 20

Could you please advise on further modifications or a better strategy to exclude linux-arm from full runs?

TG199 commented:
> > Hi @sagudev,
> > 
> > I’ve updated the try_parser to filter out the linux-arm preset when running a full run. In the handle_preset method and in the full-run section of Config.parse(), I attempted to ensure that linux-arm isn’t included in the matrix for mach try full. However, I'm still seeing the same error in GitHub Actions:
> > 
> > Invalid workflow file: .github/workflows/try.yml: too many workflows are referenced, total: 21, limit: 20
> > 
> > Could you please advise on further modifications or a better strategy to exclude linux-arm from full runs?
> 
> And this is happening when requesting just `linux-arm`? But doing manual workflow dispatch of linux-arm should work, right?

Yes it should

sagudev commented:
I think adding new entry to dispatch-workflow.yml is causing problems, so we will need to flatten some workflows (I think the best candidate for this is bencher workflow to be transformed into composite action and then called from each place respectively), but this is out of scope for this PR.

For this PR let just remove linux-arm from dispatch-workflow.yml (but keep it in try_parser even if unused) and add [`workflow_dispatch`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_dispatch) with same inputs as workflow call to be able to trigger workflow manually.

Then when we will solve conceptual problem with current CI architecture, we will be able to add linux-arm back to dispatch-workflow.yml.

TG199 commented:
> I think adding new entry to dispatch-workflow.yml is causing problems, so we will need to flatten some workflows (I think the best candidate for this is bencher workflow to be transformed into composite action and then called from each place respectively), but this is out of scope for this PR.
> 
> For this PR let just remove linux-arm from dispatch-workflow.yml (but keep it in try_parser even if unused) and add [`workflow_dispatch`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_dispatch) with same inputs as workflow call to be able to trigger workflow manually.
> 
> Then when we will solve conceptual problem with current CI architecture, we will be able to add linux-arm back to dispatch-workflow.yml.

Okay

TG199 committed:
chore: remove linux-arm from dispatch and add workflow_dispatch trigger

Signed-off-by: TG <ebiritg@gmail.com>

TG199 committed:
fix syntax error

Signed-off-by: TG <ebiritg@gmail.com>

TG199 commented:
Hi @sagudev,

I've removed the linux-arm entry from dispatch-workflow.yml as suggested, and added a workflow_dispatch trigger to linux-arm.yml with matching inputs. Now, `mach try linux-arm` works for manual runs, while linux-arm is excluded from full runs.

Could you please review these changes?

Thanks!

sagudev commented on python/servo/try_parser.py:
```diff
@@ -104,7 +107,11 @@ def update_name(self):
     s = s.lower()
 
-    if any(word in s for word in ["linux"]):
+    if s == "linux-arm":
+        return JobConfig("Linux ARM", Workflow.LINUX_ARM)
+    elif any(word in s for word in ["arm"]):
+        return JobConfig("Linux ARM", Workflow.LINUX_ARM)
```
Let's join these two:
```suggestion
    if s == "linux-arm" or any(word in s for word in ["arm"]):
        return JobConfig("Linux ARM", Workflow.LINUX_ARM)
```

sagudev committed:
Update python/servo/try_parser.py

Signed-off-by: sagudev <16504129+sagudev@users.noreply.github.com>

sagudev commented:
> Hi @sagudev,
> 
> I've removed the linux-arm entry from dispatch-workflow.yml as suggested, and added a workflow_dispatch trigger to linux-arm.yml with matching inputs. Now, `mach try linux-arm` works for manual runs, while linux-arm is excluded from full runs.
> 
> Could you please review these changes?
> 
> Thanks!

Sorry for late response. Thanks you for your work.

mrobinson commented:
Before landing this I think it makes sense that we finish this point from the TSC meeting: https://github.com/servo/project/blob/main/governance/tsc/tsc-2025-03-31.md. Let's connect on Zulip.

SuperCowProducts commented:
So... Any news?

mrobinson commented:
This was added by #46760.
